src/stores/external-source.ts

```diff
@@ -0,0 +1,185 @@
+  [derivationGroups, selectedPlanDerivationGroupNames],
+  ([$derivationGroups, $selectedPlanDerivationGroupIds]) => {
+    return $derivationGroups
+      .filter(derivationGroup => $selectedPlanDerivationGroupIds.includes(derivationGroup.name))
+      .map(derivationGroup => derivationGroup.event_types)
+      .reduce((acc, curr) => acc.concat(curr), []);
+  },
+);
+export const usersSeenSources = derived(
+  [usersSeenSourcesRaw, externalSources],
+  ([$usersSeenSourcesRaw, $externalSources]) => {
+    const res: Record<string, UserSeenEntryWithDate[]> = {};
+    for (const entry of $usersSeenSourcesRaw) {
+      const change_date = $externalSources.find(externalSource => externalSource.pkey.derivation_group_name === entry.derivation_group && externalSource.pkey.key === entry.external_source_name)?.created_at ?? ""
+      if (res[entry.username]) {
+        res[entry.username].push({
+          change_date,
+          derivation_group_name: entry.derivation_group,
+          key: entry.external_source_name,
+          source_type_name: entry.external_source_type,
+        });
+      } else {
+        res[entry.username] = [
+          {
+            change_date,
+            derivation_group_name: entry.derivation_group,
+            key: entry.external_source_name,
+            source_type_name: entry.external_source_type,
+          },
+        ];
+      }
+    }
+    return res;
+  }
+);
+
+/* Helper Functions. */
+export function resetExternalSourceStores() {
+  createExternalSourceError.set(null);
+  createExternalSourceTypeError.set(null);
+  createDerivationGroupError.set(null);
+  derivationGroupPlanLinkError.set(null);
+}
+
+function transformExternalSources(
+  externalSources: {
+    created_at: string;
+    derivation_group_name: string;
+    end_time: string;
+    key: string;
+    source_type_name: string;
+    start_time: string;
+    valid_at: string;
+  }[],
+): ExternalSourceSlim[] {
+  const completeExternalSourceSlim: ExternalSourceSlim[] = [];
+  if (externalSources !== null && externalSources !== undefined) {
+    externalSources.forEach(externalSource => {
+      completeExternalSourceSlim.push({
+        created_at: externalSource.created_at,
+        end_time: externalSource.end_time,
+        pkey: {
+          derivation_group_name: externalSource.derivation_group_name,
+          key: externalSource.key,
+        },
+        source_type_name: externalSource.source_type_name,
+        start_time: externalSource.start_time,
+        valid_at: externalSource.valid_at,
+      });
+    });
+  }
+  return completeExternalSourceSlim;
+}
+
+function transformDerivationGroups(
+  derivationGroups: {
+    derived_total: number;
+    event_types: string[];
+    name: string;
+    source_type_name: string;
+    sources: string[];
+  }[],
+): DerivationGroup[] {
+  const completeExternalSourceSlim: DerivationGroup[] = [];
+  if (derivationGroups !== null && derivationGroups !== undefined) {
```

Comment:
is this check necessary if the argument type for `derivationGroups` isn't specified as being optional?

Reply:
In case the `derivationGroups` store returns erroneously, this is just a safety check so that the store doesn't totally break. For example if the table is totally empty or doesn't exist because the migration hasn't been performed. But on second thought, that *shouldn't* ever happen, right?

Reply:
Resolved.